iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/load/cache/detector/PhiAccrualDetector.java

```diff
@@ -0,0 +1,160 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.confignode.manager.load.cache.detector;
+
+import org.apache.iotdb.confignode.manager.load.cache.AbstractHeartbeatSample;
+import org.apache.iotdb.confignode.manager.load.cache.IFailureDetector;
+import org.apache.iotdb.confignode.manager.load.cache.node.NodeHeartbeatSample;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics;
+import org.apache.tsfile.utils.Preconditions;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PhiAccrualDetector implements IFailureDetector {
+
+  private final long threshold;
+  private final long acceptableHeartbeatPauseNs;
+  private final long firstHeartbeatEstimateNs;
+  private final long minHeartbeatStdNs;
+
+  public PhiAccrualDetector(
+      long threshold,
+      long acceptableHeartbeatPauseNs,
+      long firstHeartbeatEstimateNs,
+      long minHeartbeatStdNs) {
+    this.threshold = threshold;
+    this.acceptableHeartbeatPauseNs = acceptableHeartbeatPauseNs;
+    this.firstHeartbeatEstimateNs = firstHeartbeatEstimateNs;
+    this.minHeartbeatStdNs = minHeartbeatStdNs;
+  }
+
+  @Override
+  public boolean isAvailable(List<AbstractHeartbeatSample> history) {
+    if (history.isEmpty()) {
+      /* We haven't received the first heartbeat reply. We cannot decide the node availability. */
+      return true;
+    }
+    final PhiAccrual phiAccrual = create(history);
+    // TODO(szywilliam) we should log the status change and dump the heartbeat history
+    return phiAccrual.phi() < (double) this.threshold;
+  }
+
+  PhiAccrual create(List<AbstractHeartbeatSample> history) {
+    final int size = history.size();
+
+    final List<Double> heartbeatIntervals = new ArrayList<>();
+    /*
+     * During cold start, the heartbeat history may contain not enough samples for wise decisions.
+     * Therefore, we manually add data samples for better estimation.
+     * 1. mean = heartbeat interval (1000ms)
+     * 2. std = mean / 4
+     */
+    if (size <= 2) {
+      double mean = firstHeartbeatEstimateNs;
+      double std = mean / 4.0;
+      heartbeatIntervals.add(mean + std);
+      heartbeatIntervals.add(mean - std);
+    }
+
+    for (int i = 1; i < size; i++) {
+      // ensure getSampleLogicalTimestamp() will return system nano timestamp
+      Preconditions.checkArgument(history.get(i) instanceof NodeHeartbeatSample);
+      double interval =
```

Comment:
1.It seems that as long as the duration of a heartbeat is less than 1 second, the interval should be about 1 second by default. Is this reasonable? Does a faster heartbeat predict earlier detection of problems?
2. This history is a linkedlist and cannot support get(i) requests efficiently. I recommend just creating a circular array of size 100 for slidingWindow

Reply:
Let's use iterator to avoid `get` call.